x-pack/elastic-agent/pkg/core/plugin/service/app.go

```diff
@@ -0,0 +1,330 @@
+		logger:         logger,
+		limiter:        b,
+		reporter:       reporter,
+		monitor:        monitor,
+		uid:            uid,
+		gid:            gid,
+		credsPort:      credsPort,
+	}, nil
+}
+
+// Monitor returns monitoring handler of this app.
+func (a *Application) Monitor() monitoring.Monitor {
+	return a.monitor
+}
+
+// State returns the application state.
+func (a *Application) State() state.State {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+	return a.state
+}
+
+// Name returns application name
+func (a *Application) Name() string {
+	return a.name
+}
+
+// SetState sets the status of the application.
+func (a *Application) SetState(status state.Status, msg string) {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+	a.setState(status, msg)
+}
+
+// Start starts the application with a specified config.
+func (a *Application) Start(ctx context.Context, t app.Taggable, cfg map[string]interface{}) (err error) {
+	defer func() {
+		if err != nil {
+			// inject App metadata
+			err = errors.New(err, errors.M(errors.MetaKeyAppName, a.name), errors.M(errors.MetaKeyAppName, a.id))
+		}
+	}()
+
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	cfgStr, err := yaml.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	// already started
+	if a.srvState != nil {
+		a.setState(state.Starting, "Starting")
+		a.srvState.SetStatus(proto.StateObserved_STARTING, a.state.Message)
+		a.srvState.UpdateConfig(string(cfgStr))
+	} else {
+		a.setState(state.Starting, "Starting")
+		a.srvState, err = a.srv.Register(a, string(cfgStr))
+		if err != nil {
+			return err
+		}
+	}
+
+	defer func() {
+		if err != nil {
+			if a.srvState != nil {
+				a.srvState.Destroy()
+				a.srvState = nil
+			}
+		}
+	}()
+
+	if err := a.monitor.Prepare(a.name, a.pipelineID, a.uid, a.gid); err != nil {
+		return err
+	}
+
+	if a.limiter != nil {
+		a.limiter.Add()
+	}
+
+	// start the credentials listener for the service
+	if err := a.startCredsListener(); err != nil {
+		return err
+	}
+
+	// allow the service manager to ensure that the application is started, currently this does not start/stop
+	// the actual service in the system service manager
+
+	return nil
+}
+
+// Configure configures the application with the passed configuration.
+func (a *Application) Configure(_ context.Context, config map[string]interface{}) (err error) {
+	defer func() {
+		if err != nil {
+			// inject App metadata
+			err = errors.New(err, errors.M(errors.MetaKeyAppName, a.name), errors.M(errors.MetaKeyAppName, a.id))
+		}
+	}()
+
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	if a.srvState == nil {
+		return errors.New(ErrAppNotInstalled)
+	}
+
+	cfgStr, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.New(err, errors.TypeApplication)
+	}
+	err = a.srvState.UpdateConfig(string(cfgStr))
+	if err != nil {
+		return errors.New(err, errors.TypeApplication)
+	}
+	return nil
+}
+
+// Stop stops the current application.
+func (a *Application) Stop() {
```

Comment:
Kind of suprised, that `Stop` cannot have any errors. What happens on errors on Cleanup for example?

Reply:
It can error, its just the error is handled in the Application and the error is set at that level that will propogate to the fleet manager to say that stopping failed or it was successfully stopped.